src/torchjd/autogram/_engine.py

```diff
@@ -143,46 +161,99 @@ def _hook_modules(self, modules: Iterable[nn.Module]) -> None:
+                )
 
     def compute_gramian(self, output: Tensor) -> Tensor:
-        """
-        Compute the Gramian of the Jacobian of ``output`` with respect the direct parameters of all
-        ``modules``.
-
-        :param output: The vector to differentiate. Must be a 1-D tensor.
+        r"""
+        Computes the Gramian of the Jacobian of ``output`` with respect to the direct parameters of
+        all ``modules``.
+
+        .. note::
+            This function doesn't require ``output`` to be a vector. For example, if ``output`` is
+            a matrix of shape :math:`[m_1, m_2]`, its Jacobian :math:`J` with respect to the
+            parameters will be of shape :math:`[m_1, m_2, n]`, where :math:`n` is the number of
+            parameters in the model. This is what we call a `generalized Jacobian`. The
+            corresponding Gramian :math:`G = J J^\top` will be of shape
+            :math:`[m_1, m_2, m_2, m_1]`. This is what we call a `generalized Gramian`. The number
+            of dimensions of the returned generalized Gramian will always be twice that of the
+            ``output``.
+
+            A few examples:
+                - 0D (scalar) ``output``: 0D Gramian (this can be used to efficiently compute the
+                  squared norm of the gradient of ``output``).
+                - 1D (vector) ``output``: 2D Gramian (this is the standard setting of Jacobian
+                  descent).
+                - 2D (matrix) ``output``: 4D Gramian (this can be used for :doc:`Instance-Wise
+                  Multi-Task Learning (IWMTL) <../../examples/iwmtl>`, as each sample in the batch
+                  has one loss per task).
+                - etc.
+
+        :param output: The tensor of arbitrary shape to differentiate. The shape of the returned
+            Gramian depends on the shape of this output, as explained in the note above.
         """
 
-        reshaped_output = output.reshape([-1])
+        if self._batch_dim is not None:
+            # move batched dim to the end
+            ordered_output = output.movedim(self._batch_dim, -1)
+            ordered_shape = list(ordered_output.shape)
+            batch_size = ordered_shape[-1]
+            has_non_batch_dim = len(ordered_shape) > 1
+            target_shape = [batch_size]
+        else:
+            ordered_output = output
+            ordered_shape = list(ordered_output.shape)
+            has_non_batch_dim = len(ordered_shape) > 0
+            target_shape = []
+
+        if has_non_batch_dim:
+            target_shape = [-1] + target_shape
+
+        reshaped_output = ordered_output.reshape(target_shape)
+        # There are four different cases for the shape of reshaped_output:
+        # - Not batched and not non-batched: scalar of shape []
+        # - Batched only: vector of shape [batch_size]
+        # - Non-batched only: vector of shape [dim]
+        # - Batched and non-batched: matrix of shape [dim, batch_size]
```

Comment:
```suggestion
        # - Batched and non-batched: matrix of shape [dim, batch_size]
        # If any, the batched dim is the last, and if any, the non-batched dim is the first.
```
Not mandatory, but this could help understanding the code below.